商品出品機能 #5

tanabata000 opened the pull request:
# What
商品を出品するための機能を実装し、その単体テストコードを実装
# Why
商品出品させるため。
またテストコードにより正しくバリデーションが設定されているかを確認するため。

tanabata000 committed:
正規表現書き換え

サーバー起動時に
^と$は使用できないというエラー発生
正規表現を書き換え

tanabata000 committed:
ログイン状態の場合のみ、商品出品ページへ遷移

itemモデル作成
テーブル作成
ログイン状態の場合のみ、商品出品ページへ遷移

tanabata000 committed:
商品出品機能の実装およびテストコードの実装

MVCの設定
ActiveHash導入
ActiveHashモデル生成＆商品出品画面への反映
バリデーション設定
単体テストコード実装

junyakaneko67 commented:
レビュー依頼ありがとうございます。
こちらのプルリクエストですが、実際の画面や機能ができているか確認させていただきたいので、下記キャプチャ画像をご提出いただけますか？
```
①ログイン状態の場合は、商品出品ページへ遷移できる動画
②価格が入力されると同時に、販売手数料と販売利益が表示される動画
③必要な情報を適切に入力して「出品する」ボタンを押すと、商品情報がデータベースに保存される動画
④入力に問題がある状態で「出品する」ボタンが押された場合、情報は保存されず、出品ページに戻りエラーメッセージが表示される動画
⑤ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
⑥テスト結果の画像
```
「Gyazo　使い方」とgoogleで検索し、ツールによって生成されたURLをこのPRに貼っていただければ幸いです。
その後、再度レビュー依頼フォームにてご提出をお願いいたします。

Akinori0123 commented on app/controllers/items_controller.rb:
```diff
@@ -1,6 +1,55 @@
+  before_action :move_to_index, except: [:index, :show, :search]
+
   def index
+    # 【SQLの内容】テーブル（items）選択。降順に並べ替える（投稿が新しい順）
+    query = "SELECT * FROM items ORDER BY created_at DESC"
+    # SQL使用（対象：Item 実行内容：query）
+    @items = Item.find_by_sql(query)
+    # @items = Item.order("created_at DESC")
```
**indexアクション**
indexアクション内の記述は商品一覧機能の実装となります。
現時点で実装してしまうと一覧機能ブランチで差分を作ることができなくなってしまうため、一度コメントアウトか削除しましょう。

tanabata000 replied:
修正

Akinori0123 commented on app/controllers/items_controller.rb:
```diff
@@ -1,6 +1,55 @@
 class ItemsController < ApplicationController
   # 未ログインユーザーをログインページにリダイレクトする。例外アクション[:index, :show]
   # before_action :authenticate_user!, except: [:index, :show]
+  before_action :move_to_index, except: [:index, :show, :search]
```
```
:search
```
furimaの必須機能においてはsearchアクションは使用しないため、上記は削除しましょう。

tanabata000 replied:
修正

Akinori0123 commented on app/controllers/items_controller.rb:
```diff
@@ -1,6 +1,55 @@
+    params.require(:item).permit(:image, :item_name, :item_info, :item_category_id, :item_sales_status_id, :item_shipping_fee_status_id, :item_prefecture_id, :item_scheduled_delivery_id, :item_price).merge(user_id: current_user.id)
+  end
+
+  def set_tweet
+    @tweet = Tweet.find(params[:id])
+  end
```
```
  def set_tweet
    @tweet = Tweet.find(params[:id])
  end
```
furimaではTweetモデルを使用しないため、上記は削除しましょう。

tanabata000 replied:
修正

Akinori0123 commented on app/controllers/items_controller.rb:
```diff
@@ -1,6 +1,55 @@
 class ItemsController < ApplicationController
   # 未ログインユーザーをログインページにリダイレクトする。例外アクション[:index, :show]
   # before_action :authenticate_user!, except: [:index, :show]
```
```
# before_action :authenticate_user!, except: [:index, :show]
```

> ⑤ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/2444549f43c022a2d37321b2c4861504

頂いたgyazoですとログインページに移動しておりますが、現在authenticate_user!がコメントアウトされているため上記挙動にならないかと考えられます。
改めてログアウト状態で出品ページに移動するとログインページに遷移するかどうか動作確認をしていただけますと幸いです。

tanabata000 replied:
修正

Akinori0123 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,84 @@
+        @item.item_category_id = ''
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Item category can't be blank")
+      end
```
**activehashカラムのテスト**
より厳密に要件を満たしていることを確認するために、activehashカラム全てについて以下もテストしましょう。
```
「---」が選択されている場合は出品できない
```

Akinori0123 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,84 @@
+        @item.item_price = 'aaa'
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Item price is not included in the list", "Item price is not a number")
+      end
```
**商品の単体テスト**
より厳密に要件を満たしていることを確認するために、以下もテストしましょう。
```
userが紐づいていなければ出品できない
```

Akinori0123 reviewed (commented):
gyazoのご共有ありがとうございます。確認いたしました。
数点コメントいたしましたのでご確認の程よろしくお願いいたします。

tanabata000 committed:
ログイン画面遷移

サインアウト状態で出品ボタンを押すと
ログイン画面に遷移する機能について
move_to_indexを使用した記述から
authenticate_user!を使用した記述に切り替える

tanabata000 committed:
商品出品機能（修正01）

Userテーブルを参照するための記述に変更

TakumaKawashima reviewed (commented):
レビュー依頼ありがとうございます。
LGTMとさせていただきます。実装お疲れさまでした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き頑張っていきましょう。
